src/abstractdataframe/abstractdataframe.jl

```diff
@@ -981,43 +1000,94 @@ julia> vcat(df1, df2)
 │ 4   │ 4     │ 4     │
 │ 5   │ 5     │ 5     │
 │ 6   │ 6     │ 6     │
+
+julia> vcat(df1, df3, columns = :union)
+6×3 DataFrame
+│ Row │ A     │ B       │ C       │
+│     │ Int64 │ Int64⍰  │ Int64⍰  │
+├─────┼───────┼─────────┼─────────┤
+│ 1   │ 1     │ 1       │ missing │
+│ 2   │ 2     │ 2       │ missing │
+│ 3   │ 3     │ 3       │ missing │
+│ 4   │ 7     │ missing │ 7       │
+│ 5   │ 8     │ missing │ 8       │
+│ 6   │ 9     │ missing │ 9       │
+
+julia> vcat(df1, df3, columns = :intersect)
+6×1 DataFrame
+│ Row │ A     │
+│     │ Int64 │
+├─────┼───────┤
+│ 1   │ 1     │
+│ 2   │ 2     │
+│ 3   │ 3     │
+│ 4   │ 7     │
+│ 5   │ 8     │
+│ 6   │ 9     │
+
+d = DataFrame()
+julia> vcat(d, df1)
 ```
 """
-Base.vcat(df::AbstractDataFrame) = df
-Base.vcat(dfs::AbstractDataFrame...) = _vcat(collect(dfs))
-function _vcat(dfs::AbstractVector{<:AbstractDataFrame})
+Base.vcat(dfs::AbstractDataFrame...;
+          columns::Union{Symbol, AbstractVector{Symbol}}=:same) = 
+    _vcat([df for df in dfs if (nrow(df) != 0 && ncol(df) != 0)]; columns=columns)
+
+function _vcat(dfs::AbstractVector{<:AbstractDataFrame}; 
+               columns::Union{Symbol, AbstractVector{Symbol}}=:same)
     isempty(dfs) && return DataFrame()
+    # Array of all headers
     allheaders = map(names, dfs)
+    # Array of unique headers across all data frames
     uniqueheaders = unique(allheaders)
+    # All symbols present across all headers
     unionunique = union(uniqueheaders...)
+    # List of symbols present in all dataframes
     intersectunique = intersect(uniqueheaders...)
-    coldiff = setdiff(unionunique, intersectunique)
-
-    if !isempty(coldiff)
-        # if any DataFrames are a full superset of names, skip them
-        filter!(u -> Set(u) != Set(unionunique), uniqueheaders)
-        estrings = Vector{String}(undef, length(uniqueheaders))
-        for (i, u) in enumerate(uniqueheaders)
-            matching = findall(h -> u == h, allheaders)
-            headerdiff = setdiff(coldiff, u)
-            cols = join(headerdiff, ", ", " and ")
-            args = join(matching, ", ", " and ")
-            estrings[i] = "column(s) $cols are missing from argument(s) $args"
-        end
+
+    if columns === :same
```

Comment:
If we have `:same` and we `vcat` some data frame and an empty data frame should it be OK or throw an error. (note that empty means no rows and no columns, as no-rows and some columns data frame should go a standard test)

Reply:
I don't think so. It would be nice to have some behavior like 

```
df = DataFrame()

for file in files
    d = CSV.read(file) |> DataFrame
    vcat(df, d)
end
```

We think the same behavior is helpful with `push!` and allow that. At some point I might want a PR that nails down consistency across `append!`, `vcat`, and `push!`.

Reply:
Can you please comment how is it related with `push!`? `push!` adds exactly one row. While the case we are discussing here assumes that we add 0 rows in a data frame with 0 columns.

Regarding `append!` - this is exactly the behavior we have now. I.e. you can call `append!(df, DataFrame())` and this is a no-op. So the only inconsistent function is `vcat` (where it should not be no-op, but a copy of the first `DataFrame`)

Reply:
Also you are not allowed to `push!` a `DataFrame` to a `DataFrame`.

Reply:
Oh sorry! I misread your comment! It's about `:same` rather than other methods.

I vote for `vcat(DataFrame(), df` to never throw an error. The user's behavior inside the above loop I laid out above should be dictated by the data in their files, not the rules of `vcat` with an empty data frame. 

Could you comment on the issue I raised in another comment here? The `vcat(DataFrame(), df)` methods are prohibited from tests, but its also testing behavior I don't know about.

Reply:
`DataFrame(Int, 0, 0)` is the same as `DataFrame()`.

Tests:
```
@test_throws ArgumentError vcat(missing_df, df)
@test_throws ArgumentError vcat(df, missing_df)
```
test the old behavior (disallowing `vcat`-ing non-empty `DataFrame` with empty `DataFrame`) that is exactly what I propose to change (and this is a good sign that we have tests for this in place 😄 - in general we should this level of detail in testing corner cases everywhere).